File: src/Rulesage.Shared/Repositories/Abstractions/IActionRepository.cs
using Rulesage.Common.Grammar.Ast;

namespace Rulesage.Shared.Repositories.Abstractions;

public interface IActionRepository: IDocumentRepository
{
    Task<IEnumerable<ActionExpr>> FindByIdsAsync(IEnumerable<string> ids, CancellationToken cancellationToken = default);
    
    Task<IEnumerable<(ActionExpr, float)>> FindOrderByCosineDistanceAsync(float[] queryVector, int skip, int take,
        CancellationToken cancellationToken = default);
    
    Task<bool> SaveAsync(IEnumerable<ActionExpr> actions, CancellationToken cancellationToken = default);
}

File: src/Rulesage.Common/Repositories/Abstractions/IActionRepository.fs
namespace Rulesage.Common.Repositories.Abstractions

open System.Runtime.InteropServices
open System.Threading
open System.Threading.Tasks
open Rulesage.Common.Grammar.Ast

type IActionRepository =
    inherit IDocumentRepository

    abstract FindByIdsAsync:
        ids: string seq * [<Optional; DefaultParameterValue(CancellationToken())>] cancellationToken: CancellationToken ->
            Task<ActionExpr seq>

    abstract FindOrderByCosineDistanceAsync:
        contextCommunity: string *
        queryVector: float32[] *
        skip: int *
        take: int *
        [<Optional; DefaultParameterValue(CancellationToken())>] cancellationToken: CancellationToken ->
            Task<(ActionExpr * float32) seq>

    abstract SaveAsync:
        actions: ActionExpr seq *
        [<Optional; DefaultParameterValue(CancellationToken())>] cancellationToken: CancellationToken ->
            Task<int>
